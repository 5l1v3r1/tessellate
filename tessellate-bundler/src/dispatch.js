// @flow

import logger from './logger'
import { Problem } from './error'

const log = logger('dispatch')

type DispatchArgs = { [key: string]: mixed }

type DispatchResult<K: Symbol> = { [key: string]: mixed }

type DispatchHandler<K: Symbol, A: DispatchArgs, R: DispatchResult<K>> = (args: A) => Promise<R>

type Registry = { [key: Symbol]: DispatchHandler<*, any, *> }

class DispatchProblem extends Problem {
  constructor(detail: string) {
    super({title: 'Dispatch error.', detail, status: 404})
  }
}

const registry: Registry = {}

export function register<K: Symbol, A: DispatchArgs>(name: K, handler: DispatchHandler<K, A, *>): DispatchHandler<K, A, *> {
  registry[name] = handler
  return handler
}

export default async function dispatch<K: Symbol, A: DispatchArgs>(name: K, args: A): Promise<DispatchResult<K>> {
  if (!registry[name]) {
    log.error('No such handler \'%s\'', name)
    throw new DispatchProblem(`No such handler '${name.toString()}'`)
  }
  const handler: DispatchHandler<K, A, K> = registry[name]
  return await handler(args)
}
