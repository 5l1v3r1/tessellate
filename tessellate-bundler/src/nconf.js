// @flow

import fs from 'fs'
import url from 'url'
import path from 'path'
import nconf from 'nconf'
import yaml from 'js-yaml'
import logger from './logger'

const log = logger('nconf')

function readYamlFile(file: string): Object {
  try {
    log.debug('Using configuration file %s', file)
    return yaml.safeLoad(fs.readFileSync(file, 'utf8'))
  } catch (e) {
    log.warn('Unable to load %s', file)
    return {}
  }
}

const configFile = path.resolve(process.cwd(), 'config.yaml')

export default nconf.use('memory')
                    .argv({
                      'NPM_MODULES': {
                        array: true
                      }
                    })
                    .env()
                    .add('config', {type: 'literal', store: readYamlFile(configFile)})
                    .defaults({
                      APP_PORT: 3001,
                      MORGAN_FORMAT: 'dev',
                      MORGAN_THRESHOLD: 0,
                      PUBLISH_TARGET: 'file://bundles/',
                      NPM_MODULES: [],
                      NPM_EXTERNALS: []
                    })

export function getFileSystemPublishTarget(): string | null {
  const target = url.parse(nconf.get('PUBLISH_TARGET'))
  if (target.protocol === 'file:') {
    if (!target.hostname && !target.pathname) return null
    else return path.resolve(process.cwd(), path.join(target.hostname, target.pathname))
  }
  return null
}
