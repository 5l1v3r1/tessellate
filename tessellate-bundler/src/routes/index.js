// @flow

import Router from 'koa-router'
import * as health from './health'

export type Route = (router: Router) => Router
export type RouteSet = { [key: string]: Route }

function values<T>(object: { [key: string]: T }): Array<T> {
  return Object.keys(object).map(key => object[key])
}

function initializeRoutes(router: Router, ...routes: Array<RouteSet>): Router {
  return routes.reduce((routes, set) => routes.concat(values(set)), [])
               .reduce((router, route) => route(router), router)
}

const router = initializeRoutes(new Router(), health)

export default router.routes()
