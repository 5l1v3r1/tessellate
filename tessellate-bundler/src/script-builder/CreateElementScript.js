// @flow

type CreateElementArgs = {|
  className: string;
  propsId: string;
  children: Array<string>;
|}

export default function CreateElementScript(args: CreateElementArgs): string {
  return `
  React.createElement(
    ${args.className},
    props["${args.propsId}"] ${args.children.length ? ', ' : ''}
    ${args.children.join(', ')}
  )
  `.trim().replace(/^\s+/gm, '')
}
