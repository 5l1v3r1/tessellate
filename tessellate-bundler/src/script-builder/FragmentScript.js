// @flow

type FragmentArgs = {|
  reactElement: string;
  rootID: string;
  props: { [key: string]: any };
  imports: { [key: string]: string };
|}

function ImportScript(name: string, module: string): string {
  return `import ${name} from '${module}'`
}

export default function FragmentScript(args: FragmentArgs): string {
  return `
  import React, { Component } from 'react'
  import ReactDOM from 'react-dom'
  ${Object.keys(args.imports).map(name => ImportScript(name, args.imports[name])).join('\n')}

  const BUNDLED_PROPS = ${JSON.stringify(args.props)}

  export class Fragment extends Component {
    render() {
      const props = Object.assign({}, BUNDLED_PROPS, this.props)
      return (<div id="${args.rootID}" data-props={JSON.stringify(this.props)}>
        {${args.reactElement}}
      </div>)
    }
  }

  export default function render(element) {
    const fragmentRoot = document.getElementById('${args.rootID}')
    const props = JSON.parse(fragmentRoot.getAttribute('data-props'))
    ReactDOM.render(<Fragment {...props}/>, element)
  }
  `.trim().replace(/^\s+/gm, '')
}
