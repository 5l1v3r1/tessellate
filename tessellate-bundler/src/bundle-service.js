// @flow

import path from 'path'
import webpack from 'webpack'
import ExtractTextPlugin from 'extract-text-webpack-plugin'
import WebpackSandbox from 'webpack-sandboxed'
import logger from './logger'
import nconf from './nconf'
import * as utils from './utils'

export type BundleType = {
  js: {
    name: string;
    source: string;
  };
  css?: {
    name: string;
    source: string;
  };
};

const log = logger('bundle-service')

async function _createWebpackSandbox(): Promise<WebpackSandbox> {
  const sandbox = await WebpackSandbox.createInstance({
    config: {
      target: 'web',
      module: {
        loaders: [
          {
            test: /\.js$/, exclude: /node_modules/, loader: 'babel', query: {
            presets: ['es2015', 'react']
            }
          },
          {
            test: /\.css$/, loader: ExtractTextPlugin.extract({
              fallbackLoader: 'style-loader',
              loader: 'css-loader'
            })
          }
        ]
      },
      output: {
        libraryTarget: 'umd'
      },
      plugins: [
        new webpack.DefinePlugin({process: {env: {NODE_ENV: '"production"'}}}),
        new ExtractTextPlugin('[name]-[hash].min.css')
      ].concat((nconf.get('NODE_ENV') === 'production'
          ? [
            new webpack.optimize.UglifyJsPlugin(),
            new webpack.optimize.DedupePlugin()
          ]
          : [])),
      externals: {
        'react': 'react',
        'react-dom': 'react-dom'
      }
    },
    packages: [
      'style-loader',
      'css-loader',
      'z-shop-ui' // FIXME: make configurable
    ],
    includes: [
      // path.resolve(__dirname, './components')
    ]
  })
  return sandbox
}

const createWebpackSandbox = utils.memoize(_createWebpackSandbox)

export async function make(source: string): Promise<BundleType> {
  log.debug('Create WebpackSandbox...')
  const sandbox = await createWebpackSandbox()
  log.debug('Compile bundle...')
  const [bundle, stats] = await sandbox.run(source)

  const files = Object.keys(bundle)
  log.debug('Made bundle %o', files)
  // TODO: handle case of multiple files
  const jsBundle = files.filter(b => b.endsWith('.js'))[0]
  const cssBundle = files.filter(b => b.endsWith('.css'))[0]

  const js = {
    name: jsBundle,
    source: bundle[jsBundle]
  }
  const css = cssBundle ? {
    name: cssBundle,
    source: bundle[cssBundle]
  } : undefined

  return {js, css}
}
