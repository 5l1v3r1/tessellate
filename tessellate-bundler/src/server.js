// @flow

import path from 'path'
import Koa from 'koa'
import morgan from 'koa-morgan'
import bodyParser from 'koa-bodyparser'
import koaStatic from 'koa-static'
import kcors from 'kcors'
import logger from './logger'
import routes from './routes'
import error from './error'
import nconf, { getFileSystemPublishTarget } from './nconf'
import './actions'

const log = logger('server')

export function init(): Koa {
  const app = new Koa()
  const morganFormat = nconf.get('MORGAN_FORMAT')
  const morganSkip = (req, res) => res.statusCode < nconf.get('MORGAN_THRESHOLD')

  app
    .use(morgan(morganFormat, {skip: morganSkip}))
    .use(error)
    .use(kcors())
    .use(bodyParser({enableTypes: ['json'], strict: true}))
    .use(routes)

  const fileSystemPublishTarget = getFileSystemPublishTarget()
  if (fileSystemPublishTarget) {
    log.info('Serving files from %s', fileSystemPublishTarget)
    app.use(koaStatic(fileSystemPublishTarget, {defer: true, gzip: true}))
  }

  return app
}

export function start(port: number | string = nconf.get('APP_PORT')) {
  init().listen(port)
  log.info('listening on port %d', port)
}

// $FlowIssue https://github.com/facebook/flow/issues/1362
if (require.main === module) {
  try {
    start()
  } catch(e) {
    log.error(e)
  }
}
