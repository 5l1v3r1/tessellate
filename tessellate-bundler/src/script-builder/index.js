// @flow

import uuid from 'uuid'
import logger from '../logger'
import { camelCase, constantCase } from 'change-case'
import FragmentScript from './FragmentScript'
import CreateElementScript from './CreateElementScript'
import { Problem } from '../error'

const log = logger('script-builder')

export type ElementType = {
  type: string;
  style?: string;
  props: { [key: string]: any } | null;
  children?: Array<ElementType | string>;
}

type ReactScript = {
  script: string;
  props: Object;
  imports: Object;
}

type PropsType = {[key: string]: any}
type ImportsType = {[key: string]: string}

class ElementProblem extends Problem {
  constructor(detail: string) {
    super({title: 'Element error.', detail, status: 400})
  }
}

export function build(element: ElementType): string {
  log.debug('Build fragment script...')
  const {script, props, imports} = toReactScript(element)
  return FragmentScript({reactElement: script, rootID: uuid.v4(), props, imports})
}

function toReactScript(element: ElementType | string, props: PropsType = {}, imports: ImportsType = {}): ReactScript {
  // Atomic element, e.g. string content of a headline.
  if (typeof element === 'string')
    return {script: `\`${element}\``, props, imports}

  const {className, imported} = parseElementType(element)
  const id = uuid.v4()
  props[id] = element.props

  if (imported) Object.assign(imports, imported)

  const children = toReactScripts(element.children, props, imports)

  const {childScripts, childProps} = children.reduce(({childScripts, childProps}, {script, props}) => ({
    childScripts: childScripts.concat(script),
    childProps: Object.assign(childProps, props)
  }), {childScripts: [], childProps: {}})

  const script = CreateElementScript({className, propsId: id, children: childScripts})

  return {script, props: Object.assign(props, childProps), imports}
}

function toReactScripts(elements: ?Array<ElementType | string>, props: PropsType, imports: ImportsType): Array<ReactScript> {
  if (Array.isArray(elements))
    return elements.map(child => toReactScript(child, props, imports))
  else
    return []
}

function parseElementType(element: ElementType): {|className: string; imported?: ImportsType;|} {
  if (!element.type)
    throw new ElementProblem('Missing element type on ' + JSON.stringify(element))

  // Try to match '<node-module-name>.<component-name>'
  const [_, moduleName, componentName] = element.type.match(/^([^\.]+)\.?(.+)?/) || []

  if (moduleName && componentName) {
    // Module imports for this element.
    const imported = {}
    // Add a module import based on the type.
    const importName: string = camelCase(moduleName)
    imported[importName] = moduleName
    // If a style attribute is present, add a module import for it.
    const style: ?string = constantCase(element.style)
    if (style) imported[style] = element.style
    // The React class name is '<import-name>.<component-name>'
    const className = `${importName}.${componentName}`
    return {className, imported}
  }
  // Type is a literal React component name (e.g. 'div')
  else if (moduleName) {
    const name: string = camelCase(moduleName)
    const module = moduleName
    const className = `'${name}'`
    return  {className}
  }
  else
    throw new ElementProblem(`Illegal element type ${element.type}`)
}
