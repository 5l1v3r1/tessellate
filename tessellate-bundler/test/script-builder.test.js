// @flow

jest.mock('uuid')

import fs from 'mz/fs'
import path from 'path'
import * as scriptBuilder from '../src/script-builder'

describe('script-builder', () => {
  const uuid = require('uuid')
  uuid.v4.mockImplementation(mockUUID())

  it('should fragment script string from a content structure', async () => {
    const json = await fs.readFile(path.resolve(__dirname, 'fixtures', 'content.json'))
    const element = JSON.parse(json)
    const result = scriptBuilder.build(element)
    expect(typeof result).toBe('string')
    expect(result).toMatchSnapshot()
  })
})

function mockUUID(): () => number {
  let i = 0
  return () => ++i
}
