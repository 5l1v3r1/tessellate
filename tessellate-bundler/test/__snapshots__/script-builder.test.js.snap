exports[`script-builder should fragment script string from a content structure 1`] = `
"import React, { Component } from \'react\'
import ReactDOM from \'react-dom\'
import mosaicComponentExampleDistStylesCss from \'mosaic-component-example/dist/styles.css\'
import mosaicComponentExample from \'mosaic-component-example\'
const BUNDLED_PROPS = {\"1\":null,\"2\":{\"level\":3}}
export class Fragment extends Component {
render() {
const props = Object.assign({}, BUNDLED_PROPS, this.props)
return (<div id=\"3\" data-props={JSON.stringify(this.props)}>
{React.createElement(
\'div\',
props[\"1\"] , 
React.createElement(
mosaicComponentExample,
props[\"2\"] , 
\`Hello, World!\`
)
)}
</div>)
}
}
export default function render(element) {
const fragmentRoot = document.getElementById(\'3\')
const props = JSON.parse(fragmentRoot.getAttribute(\'data-props\'))
ReactDOM.render(<Fragment {...props}/>, element)
}"
`;
